### PR TITLE
DBZ-217 Handling exceptions ocurring during binlog event deserialization

### PR DESCRIPTION
**Pull request by gunnarmorling**
* Catching the exception and wrapping it into a pseudo INCIDENT event
* Logging the exception and the current SourceInfo state

https://issues.jboss.org/browse/DBZ-217

@rhauch I'm not yet familiar with the binlog format, but I hope the source info provides the state required for spotting the flawed event? It'll log output like so:

    2017-05-31 16:16:28,909 ERROR  MySQL|myServer1|binlog  Exception during event data deserialization (source info: binlog file 'mysql-bin.000002', pos=879, skipping 12758 events plus 0 rows, partion: {server=myServer1})   [io.debezium.connector.mysql.BinlogReader]
    com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1496225416000, eventType=EXT_WRITE_ROWS, serverId=112233, headerLength=19, dataLength=40, nextPosition=2662649, flags=0}
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:212)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:180)
        at io.debezium.connector.mysql.BinlogReader$1.nextEvent(BinlogReader.java:120)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:741)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:472)
        at com.github.shyiko.mysql.binlog.BinaryLogClient$5.run(BinaryLogClient.java:657)
        at java.lang.Thread.run(Thread.java:745)
    Caused by: java.io.EOFException
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:190)
        at java.io.InputStream.read(InputStream.java:170)
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.fill(ByteArrayInputStream.java:96)
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:89)
        at io.debezium.connector.mysql.RowDeserializers.deserializeString(RowDeserializers.java:246)
        at io.debezium.connector.mysql.RowDeserializers$WriteRowsDeserializer.deserializeString(RowDeserializers.java:185)
        at io.debezium.connector.mysql.BinlogReader$MyWriteRowsDeserializer.deserializeString(BinlogReader.java:174)
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeCell(AbstractRowsEventDataDeserializer.java:176)
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeRow(AbstractRowsEventDataDeserializer.java:132)
        at com.github.shyiko.mysql.binlog.event.deserialization.WriteRowsEventDataDeserializer.deserializeRows(WriteRowsEventDataDeserializer.java:64)
        at com.github.shyiko.mysql.binlog.event.deserialization.WriteRowsEventDataDeserializer.deserialize(WriteRowsEventDataDeserializer.java:56)
        at com.github.shyiko.mysql.binlog.event.deserialization.WriteRowsEventDataDeserializer.deserialize(WriteRowsEventDataDeserializer.java:32)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:206)
        ... 6 more

Automating a test for this is very hard but I tested it manually by plugging in a custom deserializer which would provoke an EOF exception and I got the output above. Note the exception is just logged, not re-thrown, i.e. it will continue to read events after that (in the case of EOF there may not be more to read of course).

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 1]
```diff
@@ -374,18 +385,28 @@ protected void handleServerHeartbeat(Event event) {
...
      * @param event the server stopped event to be processed; may not be null
      */
     protected void handleServerIncident(Event event) {
-        logger.trace("Server incident: {}", event);
+        if (event.getData() instanceof EventDataDeserializationExceptionData) {
+            EventDataDeserializationExceptionData data = event.getData();
+
+            logger.error(
+                    String.format("Exception during event data deserialization (source info: %s, partion: %s)", source, source.partition()),
+                    data.getCause()
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: I think we need to be much more clear that this event was skipped.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 2]
```diff
@@ -112,14 +113,24 @@ public BinlogReader(String name, MySqlTaskContext context) {
...
+                        tableMapEventByTableId.put(tableMapEvent.getTableId(), tableMapEvent);
+                    }
+                    return event;
+                }
+                // DBZ-217 In case an event couldn't be read we create a pseudo-event for the sake of logging
+                catch(EventDataDeserializationException edde) {
+                    EventHeaderV4 header = new EventHeaderV4();
+                    header.setEventType(EventType.INCIDENT);
+                    EventData data = new EventDataDeserializationExceptionData(edde);
+                    return new Event(header, data);
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: IIRC, we can get this exception for other cases, like unknown MySQL event types. That would be rare, but I'm wondering if we should constrain this as much as possible. What are the potential cases other than the corrupt binlog even where this might be thrown?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 3]
```diff
@@ -374,18 +385,28 @@ protected void handleServerHeartbeat(Event event) {
...
+        if (event.getData() instanceof EventDataDeserializationExceptionData) {
+            EventDataDeserializationExceptionData data = event.getData();
+
+            logger.error(
+                    String.format("Exception during event data deserialization (source info: %s, partion: %s)", source, source.partition()),
+                    data.getCause()
+            );
+        }
+        else {
+            logger.warn("Server incident: {}", event);
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: What other kinds of server incidents are possible? Should this log them at ERROR rather than WARN?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 1 above)
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: In fact, it'd be great if we actually wrote out what the user should do to see more info about this event, such as running something like the following to see this particular event:
```
mysqlbinlog --start-position=1200 --stop-position=1201 --verbose binlog.0000003
```
where the start position is access from `source.restartBinlogPosition()`, stop position is obtained from `source.nextBinlogPosition()` that needs to be added (and should return `source.currentBinlogPosition + source.currentEventLengthInBytes`), and the binlog filename from `source.binlogFilename()`. I think these are correct, since they are set [here](https://github.com/gunnarmorling/debezium/blob/6a8add739f77a43ea18659c1e992cc8a11d4344d/debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java#L324) before the [event is handled](https://github.com/gunnarmorling/debezium/blob/6a8add739f77a43ea18659c1e992cc8a11d4344d/debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java#L330) and thus before we might process this exception.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 4]
```diff
@@ -374,18 +385,28 @@ protected void handleServerHeartbeat(Event event) {
...
+     *
      * @param event the server stopped event to be processed; may not be null
      */
     protected void handleServerIncident(Event event) {
-        logger.trace("Server incident: {}", event);
+        if (event.getData() instanceof EventDataDeserializationExceptionData) {
+            EventDataDeserializationExceptionData data = event.getData();
+
+            logger.error(
+                    String.format("Exception during event data deserialization (source info: %s, partion: %s)", source, source.partition()),
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Why are you using `String.format` here? The `error(...)` method takes parameters like `{}`. So in this case, you'd do something like:
```
  logger.error("Error while deserializing binlog event at {}", source.offset(), data.getCause());
```
Note how the exception is 1 more argument than the last parameter value.

But as I suggest in my next comment, it'd be good to also mention in this log message how to view the event with `mysqlbin`, so:
```
  logger.error("Error while deserializing binlog event at {}\nUse the `mysqlbinlog` tool to view the problematic event with `mysqlbinlog --start-position={} --stop-position={} --verbose {}", 
           source.offset(), source.binlogPosition(), source.nextBinlogPosition(), source.binlogFilename(), data.getCause());
```

Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: May require reformatting/rewording the log message to make it easier to read.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 2 above)
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: This exception is raised only when something goes wrong during deserialization of event data. If we come across an unknown event it's my understanding that it will be ignored.

Or is it that you are saying the try/catch here should be _more_ general and not only handle `EventDataDeserializationException`?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 4 above)
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: > Note how the exception is 1 more argument than the last parameter value.

Ah, that's nice, wasn't aware of this special usage. I'll change it.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 3 above)
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Yes, I'll change it to ERROR. It used to be TRACE before which didn't seem right, so I changed it to WARN. But ERROR is better indeed. As to when it can happen, the docs say that: "something happened on the master that might cause data to be in an inconsistent state". But we'll handle them as before (just logging), only our special data will cause the special logging with more details.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 2 above)
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: My concern is that `EventDataDeserializationException` is thrown anytime the binlog client library has an error reading the input stream, which will include situations where the binlog client library has a bug and is trying to read the event incorrectly. If that were the case, then this code would eat and log that event rather than failing.

The situation that we're trying to handle is a subset of cases where `EventDataDeserializationException` is thrown, and specifically the case where the binlog event is actually corrupted. Unfortunately, I don't see how we can discover that subset case, so I'm not sure we can do anything but what you've coded.

Bottom line is that I'm worried this change will make the connector eat/log more errors than it should, at which point those events are now lost and any data constructed using our event stream is potentially corrupt.

**Comment by gunnarmorling** [comment]
> My concern is that EventDataDeserializationException is thrown anytime
the binlog client library has an error reading the input stream

Do you have any pointer to that? The only place I can find where it's
raised is in
com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData().
So it seemed quite save to me.

**Comment by rhauch** [comment]
Yes, that's the spot. Any `IOException` thrown in [this block|https://github.com/shyiko/mysql-binlog-connector-java/blob/604079d76fa373b07cfa1914ccc6b7cf116d402b/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/EventDeserializer.java#L216] will result in an `EventDataDeserializationException`.

The only time this is an issue is if the binlog client code and the event don't agree on the format.
I agree that _theoretically_ the binlog client code is correct and this should only happen with corrupted events, but it will also happen if there is a bug in the binlog client code.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 2 above)
Comment (jpechane) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: I agree it is a potentially risky behaviour so I think we should do two things to guard against problem

- Make the behaviour disabled by default and provide configuration option to enable it
- I've already proposed it in chat. Instead of just informing the user and throwing the event we should forward it into a special topic (like DLQ in messaging). This could be monitored by an operator and appropriate actipons can be executed.

**Comment by rhauch** [comment]
I think we'll need to add an option that will say how to handle such errors. I can very easily see users want to fail on every error, while others might not consider these to be serious would rather have missed events than stop their connector.

**Comment by gunnarmorling** [comment]
Agreed, an option would make sense very much.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 3 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Done.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: I've changed the message as suggested. It'll say:

    Error while deserializing binlog event at offset {ts_sec=1497445917, file=mysql-bin.000002, pos=3006051, server_id=223344, event=2}.
    Use the mysqlbinlog tool to view the problematic event: `mysqlbinlog --start-position=3006194 --stop-position=3006379 --verbose mysql-bin.000002

There's an option that controls what to do in case of a deserialization error: ignore it, warn about it and proceed or stop further processing. In the second case, the message above will also contain this note:

    "This exception will be ignored and the event be skipped.

Otherwise, the exception will be raised and processing stops.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 4 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Done.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: It's an option now, with the default being the current behavior, i.e. stopping the connector, but logging the context required to analyze the event in the binlog. I think the DLQ approach is nice, but something we can handle separately and add it as further value of that new option.

**Comment by gunnarmorling** [comment]
@rhauch could you take another look? The option to control the behavior is still missing, but I was hoping I could get your feedback on the actual change in the binlog reader.

It foresees to examine that new option, with the current behavior being the default (stopping the connector) and optionally allowing to skip any corrupt event. In either case, the event context and the exception will be logged.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 5]
```diff
@@ -378,18 +399,54 @@ protected void handleServerHeartbeat(Event event) {
...
+            // TODO make it an option
+            EventDeserializationFailureHandlingMode deserializationFailureHandling = EventDeserializationFailureHandlingMode.FAIL;
+            EventHeaderV4 eventHeader = (EventHeaderV4) data.getCause().getEventHeader(); // safe cast, instantiated that ourselves
+
+            if(deserializationFailureHandling == EventDeserializationFailureHandlingMode.FAIL) {
+                // logging some additional context but not the exception itself, this will happen in handleEvent()
+
+                logger.error(
+                        "Error while deserializing binlog event at offset {}." + System.lineSeparator() +
+                        "Use the mysqlbinlog tool to view the problematic event: `mysqlbinlog --start-position={} --stop-position={} --verbose {}",
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Why not use a parameter for `System.lineSeparator()` so that we don't have to do string manipulation here and so that the message is clearer?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 6]
```diff
@@ -378,18 +399,54 @@ protected void handleServerHeartbeat(Event event) {
...
+                        source.binlogFilename()
+                );
+
+                throw new RuntimeException(data.getCause());
+            }
+            else if(deserializationFailureHandling == EventDeserializationFailureHandlingMode.WARN) {
+                logger.warn(
+                        "Error while deserializing binlog event at offset {}." + System.lineSeparator() +
+                        "This exception will be ignored and the event be skipped." + System.lineSeparator() +
+                        "Use the mysqlbinlog tool to view the problematic event: `mysqlbinlog --start-position={} --stop-position={} --verbose {}",
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Again, same thing here.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 7]
```diff
@@ -378,18 +399,54 @@ protected void handleServerHeartbeat(Event event) {
...
+                        source.offset(),
+                        eventHeader.getPosition(),
+                        eventHeader.getNextPosition(),
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Won't this show the *next* binlog event that is after the problematic one?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 8]
```diff
@@ -357,6 +357,21 @@ public static SecureConnectionMode parse(String value, String defaultValue) {
...
+            WARN,
+            FAIL;
+
+            @Override
+            public String getValue() {
+                return null;
+            }
+
+
+
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Minor: do we need all these blank lines? :-)

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 9]
```diff
@@ -378,18 +399,54 @@ protected void handleServerHeartbeat(Event event) {
...
+     *
      * @param event the server stopped event to be processed; may not be null
      */
     protected void handleServerIncident(Event event) {
-        logger.trace("Server incident: {}", event);
+        if (event.getData() instanceof EventDataDeserializationExceptionData) {
+            EventDataDeserializationExceptionData data = event.getData();
+
+            // TODO make it an option
+            EventDeserializationFailureHandlingMode deserializationFailureHandling = EventDeserializationFailureHandlingMode.FAIL;
```
Comment (rhauch) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: `deserializationFailureHandling` is always `FAIL`... are we missing where that's actually set somewhere? Should it be set via a new configuration property?

Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Yes, that's what I meant by

> The option to control the behavior is still missing

It will be set via that new option.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 5 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Ok, good idea.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 7 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Hum, no, I don't think so. It uses getPosition() as the start and and nextPosition() as the stop position. Copying the resulting output and invoking _mysqlbinlog_ pointed me to the right offset.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java (hunk 8 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Nope :) That's just a placeholder, still need to code that option and parsing of the enum. Just wanted to make sure you agree on the meat of the change, i.e. the logic within the binlog reader. 

Thanks a lot for the quick feedback!

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 10]
```diff
@@ -773,7 +832,9 @@ public static final Field MASK_COLUMN(int length) {
                                                      GTID_SOURCE_FILTER_DML_EVENTS,
                                                      TIME_PRECISION_MODE, DECIMAL_HANDLING_MODE,
                                                      SSL_MODE, SSL_KEYSTORE, SSL_KEYSTORE_PASSWORD,
-                                                     SSL_TRUSTSTORE, SSL_TRUSTSTORE_PASSWORD, JDBC_DRIVER);
+                                                     SSL_TRUSTSTORE, SSL_TRUSTSTORE_PASSWORD, JDBC_DRIVER,
+                                                     BIGINT_UNSIGNED_HANDLING_MODE,
```
Comment (jpechane) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Does it belong to this PR?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 11]
```diff
@@ -797,9 +858,11 @@ protected static ConfigDef configDef() {
                     DatabaseHistory.SKIP_UNPARSEABLE_DDL_STATEMENTS);
         Field.group(config, "Events", INCLUDE_SCHEMA_CHANGES, TABLES_IGNORE_BUILTIN, DATABASE_WHITELIST, TABLE_WHITELIST,
                     COLUMN_BLACKLIST, TABLE_BLACKLIST, DATABASE_BLACKLIST,
-                    GTID_SOURCE_INCLUDES, GTID_SOURCE_EXCLUDES, GTID_SOURCE_FILTER_DML_EVENTS);
+                    GTID_SOURCE_INCLUDES, GTID_SOURCE_EXCLUDES, GTID_SOURCE_FILTER_DML_EVENTS,
+                    EVENT_DESERIALIZATION_FAILURE_HANDLING_MODE);
         Field.group(config, "Connector", CONNECTION_TIMEOUT_MS, KEEP_ALIVE, MAX_QUEUE_SIZE, MAX_BATCH_SIZE, POLL_INTERVAL_MS,
-                    SNAPSHOT_MODE, SNAPSHOT_MINIMAL_LOCKING, TIME_PRECISION_MODE, DECIMAL_HANDLING_MODE);
+                    SNAPSHOT_MODE, SNAPSHOT_MINIMAL_LOCKING, TIME_PRECISION_MODE, DECIMAL_HANDLING_MODE,
+                    BIGINT_UNSIGNED_HANDLING_MODE);
```
Comment (jpechane) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Does it belong to this PR?

**Comment by gunnarmorling** [comment]
> Does it belong to this PR?

Not strictly. It was forgotten before and it seemed easy enough to just add
it while adding the other property.

**Comment by jpechane** [comment]
I undertand, ok then, I've just wanted to make sure it is intentional.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 12]
```diff
@@ -378,18 +403,54 @@ protected void handleServerHeartbeat(Event event) {
...
+                        "Error while deserializing binlog event at offset {}.{}" +
+                        "Use the mysqlbinlog tool to view the problematic event: mysqlbinlog --start-position={} --stop-position={} --verbose {}",
+                        source.offset(),
+                        System.lineSeparator(),
+                        eventHeader.getPosition(),
+                        eventHeader.getNextPosition(),
+                        source.binlogFilename()
+                );
+
+                throw new RuntimeException(data.getCause());
```
Comment (jpechane) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Should not be this CnnectException?

**Comment by gunnarmorling** [comment]
@jpechane Successfully tested it myself now. Also rebased and pushed one more commit with a small fix. Can you take a last look and merge to master and 0.6 if you think it's alright? Thanks!

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 5 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Done now.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java (hunk 9 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: The option is there now, defaulting to `FAIL`, i.e. the current behavior.

**Comment by jpechane** [comment]
Applied to master and 0.6 - thanks!
